### PR TITLE
Cleanup Empty AutoAliasing Objects

### PR DESCRIPTION
**Pull request by iwahbe**
This PR modifies token mapping to not add extra fields for otherwise empty objects. This is necessary because `AutoName` sets a full `tfbridge.SchemaInfo`, it does not modify one.

### CODE REVIEW

**Comment by t0yv0** [comment]
How does this connect to the user problem prompting this PR, would it be possible to elaborate? The tests are at a level of indirection, no elucidating new tests are added,  and it's not obvious how the problem is being fixed.

```
I believe you will recover old behavior by putting p.SetAutonaming before the call to x.AutoAliasing.
```

After this PR do we believe that the order of p.SetAutonaming and x.AutoAliasing does not matter anymore? 

Does editing the existing test not weaken the MaxItems one test strength?

I'm going to be able to review tomorrow morning but it's not an obvious fix to review in a tired state.

File: pkg/tfbridge/x/tokens.go [hunk 1]
```diff
@@ -649,62 +649,96 @@ func aliasResource(
...
+
 			return true
 		})
+
+		return rHasH, rHasI
 	}
 
-	var walk func(*fieldHistory, *b.SchemaInfo, shim.Schema)
-	walk = func(h *fieldHistory, info *b.SchemaInfo, schema shim.Schema) {
+	walk = func(h *fieldHistory, info *b.SchemaInfo, schema shim.Schema) (hasH bool, hasI bool) {
```
Comment (guineveresaenger) [inline pkg/tfbridge/x/tokens.go:688]: I think there is a lot of state context here that gets confusing because the variable names are so similar.
For example, we return `hasH` and `hasI` here but on line 667 we assign them to `keepH` and `keepI` and it would be great to know what is prompting us to "keep" info or history.

File: pkg/tfbridge/x/tokens.go [hunk 2]
```diff
@@ -649,62 +649,96 @@ func aliasResource(
...
+			return hist != nil, info != nil
 		}
+
+		var rHasH, rHasI bool
+
 		m.Range(func(k string, v shim.Schema) bool {
-			f(k, v)
+			h, hasH := getNonNil(hist, k)
+			i, hasI := getNonNil(info, k)
+			keepH, keepI := walk(h, i, v)
```
Comment (guineveresaenger) [inline pkg/tfbridge/x/tokens.go:667]: I'm not sure what is being kept here.

_Should_ we keep history/info?
Are we checking _whether_ history or info are being kept?

File: pkg/tfbridge/x/tokens.go [hunk 3]
```diff
@@ -649,62 +649,96 @@ func aliasResource(
...
+
+			hasH = hasH || keepH
+			hasI = hasI || keepI
+
+			if !hasH {
+				delete(*hist, k)
```
Comment (guineveresaenger) [inline pkg/tfbridge/x/tokens.go:673]: This deletes both the type/field key and its content, correct?

File: pkg/tfbridge/x/tokens.go [hunk 4]
```diff
@@ -713,27 +747,24 @@ func aliasResource(
...
-			getNonNil(&hist[tfToken].Fields, k),
-			getNonNil(&computed.Fields, k),
-			v,
-		)
-	})
+	walkResource(res, &hist[tfToken].Fields, &computed.Fields)
 }
 
-func getNonNil[K comparable, V any](m *map[K]*V, key K) *V {
+func getNonNil[K comparable, V any](m *map[K]*V, key K) (_ *V, alreadyThere bool) {
```
Comment (guineveresaenger) [inline pkg/tfbridge/x/tokens.go:753]: I'm not familiar with this Go pattern, and I could not find any quick reference to it - why are we returning a blank identifier here?

File: pkg/tfbridge/x/tokens.go [hunk 5]
```diff
@@ -713,27 +747,24 @@ func aliasResource(
...
 		*m = map[K]*V{}
 	}
 	v := (*m)[key]
+
 	if v == nil {
 		var new V
 		v = &new
 		(*m)[key] = v
+	} else {
+		alreadyThere = true
```
Comment (guineveresaenger) [inline pkg/tfbridge/x/tokens.go:765]: Where is `alreadyThere` initialized?

**Review by guineveresaenger — COMMENTED** [review]
I would like some comments that describe the recursion pattern, and the purpose of each inline function.
Can we also link the Docker PR that failed, with an explanation of why that happened?

Thank you for the quick turnaround.

File: pkg/tfbridge/x/tokens.go (hunk 3 above)
Comment (iwahbe) [reply pkg/tfbridge/x/tokens.go:673]: This removes the map entry indexed by the key `k` (`string) from the map `*hist` (`map[string]*fieldHistory`).

Did that answer your question?

File: pkg/tfbridge/x/tokens.go (hunk 4 above)
Comment (iwahbe) [reply pkg/tfbridge/x/tokens.go:753]: The feature I'm using is called [Named return values
](https://go.dev/tour/basics/7). I wanted to name the second returned value, since I often get these types of bool indicators confused. I didn't feel the need to name the first returned value (type `*V`). `_` is a special identifier in Go that can never be referenced.[^1]

[^1]: If you are familiar with lisp, you can conceptually replace each use of `_` with `(gensym)`.

File: pkg/tfbridge/x/tokens.go (hunk 5 above)
Comment (iwahbe) [reply pkg/tfbridge/x/tokens.go:765]: As part of the [named return value](https://go.dev/tour/basics/7).

**Comment by iwahbe** [comment]
> How does this connect to the user problem prompting this PR, would it be possible to elaborate? The tests are at a level of indirection, no elucidating new tests are added, and it's not obvious how the problem is being fixed.
> 
> ```
> I believe you will recover old behavior by putting p.SetAutonaming before the call to x.AutoAliasing.
> ```
> 
> After this PR do we believe that the order of p.SetAutonaming and x.AutoAliasing does not matter anymore?
> 
> Does editing the existing test not weaken the MaxItems one test strength?
> 
> I'm going to be able to review tomorrow morning but it's not an obvious fix to review in a tired state.

After this PR we do believe that `p.SetAutonaming` and `x.AutoAliasing` should be callable in any order.[^1] We want this property to avoid confusion.

The test was more specific than necessary. Specifically, it was testing for a property that was not necessary for the desired behavior that we want to change:

```patch
- assert.Nil(t, info.Resources["pkg_r1"].Fields["f2"].MaxItemsOne)
+ assert.Nil(t, info.Resources["pkg_r1"].Fields["f2"])
```

Previously, `info.Resources["pkg_r1"].Fields["f2"]` was equal to `&tfbridge.SchemaInfo{}`, and we only cared that `info.Resources["pkg_r1"].Fields["f2"].MaxItemsOne` was `nil`. We are tightening the requirements to ensure that the full thing is nil. 

What we actually cared about is this property (expressed in TS style pseudo-code):
```
info?.Resources?["pkg_r1"]?.Fields?["f2"]?.MaxItemsOne == nil
```

Because of how painful it is to express this in go, I abbreviated to asserting `MaxItemsOne` was `nil`.

[^1]: I believe that `SetAutonaming` will break at runtime when it is applied to a non-string field, so I'm discounting the possibility of a list type with auto-naming called `name`.

File: pkg/tfbridge/x/token_test.go [hunk 6]
```diff
@@ -746,9 +747,98 @@ func TestMaxItemsOneAliasingNested(t *testing.T) {
 	assert.False(t, *info.Resources["pkg_r1"].Fields["f2"].Elem.Fields["n2"].MaxItemsOne)
 }
 
+// (*ProviderInfo).SetAutonaming only applies to fields that are not present in their
+// resource's property map. We need to make sure that unless we mark a field as
+// `MaxItemsOne: nonNil` for some non-nil value, we don't leave that field entry behind
+// since that will disable SetAutonaming.
+func TestMaxItemsOneAliasingWithAutoNaming(t *testing.T) {
```
Comment (iwahbe) [inline pkg/tfbridge/x/token_test.go:754]: @guineveresaenger @t0yv0 Sorry for the confusing PR. I have added our motivating example as a test case with an explanation of the underlying logic for the change.

File: pkg/tfbridge/x/token_test.go [hunk 7]
```diff
@@ -514,6 +514,7 @@ func TestMaxItemsOneAliasing(t *testing.T) {
 					"pkg_r1": (&schema.Resource{Schema: schema.SchemaMap{
 						"f1": Schema{MaxItemsOne: f1},
 						"f2": Schema{MaxItemsOne: f2},
+						"f3": Schema{typ: shim.TypeString},
```
[the remark this answers is not in the document]
Comment (iwahbe) [reply pkg/tfbridge/x/token_test.go:517]: It was intentional. I wanted to test that we don't record the `MaxItemsOne` state for a field where the concept doesn't apply (type string).

File: pkg/tfbridge/x/tokens.go (hunk 2 above)
Comment (iwahbe) [reply pkg/tfbridge/x/tokens.go]: We should keep it when (1) it already existed or (2) we stored useful information in it. I added some comments explaining what we are trying to do that will hopefully clarify.

File: pkg/tfbridge/x/tokens.go (hunk 1 above)
Comment (iwahbe) [reply pkg/tfbridge/x/tokens.go]: I added that comment.

**Comment by t0yv0** [comment]
> Because of how painful it is to express this in go, I abbreviated to asserting MaxItemsOne was nil.

Yes that's true. Yes. I think we try not to care anywhere about the different flavors of "missing", not to attach semantics to nil vs missing key here, so what could work is getter functions that isolate the caller from nils, like this:

```
assert.Nil(t, info.Resource("pkg_r1").Field("f2").MaxItemsOne)
```

But perhaps for another day.

File: pkg/tfbridge/x/tokens.go (hunk 5 above)
Comment (t0yv0) [reply pkg/tfbridge/x/tokens.go:817]: I use named return values a bit. They do have a place it seems.

File: pkg/tfbridge/x/token_test.go (hunk 6 above)
Comment (t0yv0) [reply pkg/tfbridge/x/token_test.go:754]: Brilliant. Oof, so  SetAutonaming does look at whether the SchemaInfo is present or no.

Saw this code:

```go
				if _, hasfield := res.Fields[nameProperty]; !hasfield {
					if res.Fields == nil {
						res.Fields = make(map[string]*SchemaInfo)
					}
					res.Fields[nameProperty] = AutoName(nameProperty, maxLength, separator)
				}
```

I would rephrase

> (*ProviderInfo).SetAutonaming only applies to fields that are not present in their
 // resource's property map.

With 

> (*ProviderInfo).SetAutonaming skips fields that have a SchemaInfo already defined in ResourceInfo.Fields


I always confuse the Info things with he provider maps from shim.SchemaMap.

File: pkg/tfbridge/x/tokens.go [hunk 8]
```diff
@@ -650,61 +650,147 @@ func aliasResource(
...
+		populateElem := func() {
 			if h.Elem == nil {
 				h.Elem = &fieldHistory{}
+			} else {
+				hasElemH = true
 			}
 			if info.Elem == nil {
 				info.Elem = &b.SchemaInfo{}
+			} else {
+				hasElemH = true
```
Comment (t0yv0) [inline pkg/tfbridge/x/tokens.go:763]: Is this hasElemH or should it be hasElemI?

**Comment by t0yv0** [comment]
Well this is interesting, it works around the immediate problem but looks a tad error prone. I'm a bit worried about this quirk we found in SetAutonaming behavior, and perhaps there's still conditions where the order of application matters.

Consider this:

- autoaliasing runs and injects a MaxItems=1 specification, creating a SchemaInfo in ResourceInfo.Fields
- then autonaming runs and finds that whoops, the user set ResourceInfo.Fields["spock"] so I'm going to skip it

To the end user this looks like we accidentally disabled auto naming.
  
Could we do better? Could we enforce the order so SetAutonaming is always before aliasing?

**Review by jazzyfresh — APPROVED** [review]
LGTM, especially with the new test

**Comment by iwahbe** [comment]
> Well this is interesting, it works around the immediate problem but looks a tad error prone. I'm a bit worried about this quirk we found in SetAutonaming behavior, and perhaps there's still conditions where the order of application matters.


Unfortunately, I think the quirk is here to stay. I've seen plenty of `tk: { Fields: map[string]*tfbridge.SchemaInfo{ "name": {Name: "name" } } }`. AFAIK, this exists only to disable auto-naming. 

AutoNaming should have been more selective by checking if `*SchemaInfo.Defaults` was `nil` instead of `*SchemaInfo`, but I think that change needs to wait for bridge v4.0.0.

> 
> Consider this:
> 
> * autoaliasing runs and injects a MaxItems=1 specification, creating a SchemaInfo in ResourceInfo.Fields
> * then autonaming runs and finds that whoops, the user set ResourceInfo.Fields["spock"] so I'm going to skip it
> 
> To the end user this looks like we accidentally disabled auto naming.

If `AutoAliasing` saw a resource and injected `ResourceInfo["spock"].MaxItemsOne = True()`, then `SetAutonames` would not be affected, since it only effects fields called `"name"`.

If `AutoAliasing` saw a resource and injected `ResourceInfo["name"].MaxItemsOne = True()`, it would imply that `ResourceMap.Get(resourceName).Schema().Get("name").Type` is [either a `List`, `Map` or `Set` ](https://github.com/pulumi/pulumi-terraform-bridge/blob/7747fa11af1899be0b17ec3889662972b2442567/pkg/tfbridge/x/tokens.go#L732)(something that `MaxItemsOne` makes sense on). If that is the case, then `SetAutonaming` would be invalid, since it assumes that the name field is of type `String`.

The end user will never observe `AutoAliasing` disabling a valid auto name.

> Could we do better? Could we enforce the order so SetAutonaming is always before aliasing?

We could enforce the ordering, but I don't think we should.
1. We impose accidental complexity on our users. Since we can't order enforce in the type system, this is a runtime panic when developing a provider. We should avoid this where possible.
2. This cuts down the size of `bridge-metadata.json` to only keep the data it needs. This makes diffs much more meaningful.

**Comment by t0yv0** [comment]
Brilliant, so the cases don't overlap based on type. Autonaming is interested in name type=string and maxitems=1 aliasing is looking at type=list[x] or type=set[x]. That's great. 

> AutoNaming should have been more selective by checking if *SchemaInfo.Defaults was nil instead of *SchemaInfo, but I think that change needs to wait for bridge v4.0.0.

I like that. Care to backlog and label say impact/breaking? We can then find all those things when contemplating 4.0.

**Review by t0yv0 — APPROVED** [review]

**Review by guineveresaenger — APPROVED** [review]
thank you for the extra comments and improved varnames! 💟

**Comment by iwahbe** [comment]
> Brilliant, so the cases don't overlap based on type. Autonaming is interested in name type=string and maxitems=1 aliasing is looking at type=list[x] or type=set[x]. That's great.
> 
> > AutoNaming should have been more selective by checking if *SchemaInfo.Defaults was nil instead of *SchemaInfo, but I think that change needs to wait for bridge v4.0.0.
> 
> I like that. Care to backlog and label say impact/breaking? We can then find all those things when contemplating 4.0.

Follow up issue is https://github.com/pulumi/pulumi-terraform-bridge/issues/1069.
